File: backend/internal/services/notification_service.go
//go:build exclude

package services

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"strconv"
	"time"

	firebase "firebase.google.com/go"
	"firebase.google.com/go/messaging"
	"thyne-jewels-backend/internal/models"
	"thyne-jewels-backend/internal/repository"
	"go.mongodb.org/mongo-driver/bson/primitive"
	"google.golang.org/api/option"
)

// NotificationService handles push notifications
type NotificationService struct {
	notificationRepo repository.NotificationRepository
	userRepo         repository.UserRepository
	fcmClient        *messaging.Client
	emailService     *EmailService
	smsService       *SMSService
}

// NewNotificationService creates a new notification service
func NewNotificationService(
	notificationRepo repository.NotificationRepository,
	userRepo repository.UserRepository,
	firebaseCredentialsPath string,
	emailService *EmailService,
	smsService *SMSService,
) (*NotificationService, error) {
	// Initialize Firebase app
	opt := option.WithCredentialsFile(firebaseCredentialsPath)
	app, err := firebase.NewApp(context.Background(), nil, opt)
	if err != nil {
		return nil, fmt.Errorf("failed to initialize Firebase app: %w", err)
	}

	// Get messaging client
	fcmClient, err := app.Messaging(context.Background())
	if err != nil {
		return nil, fmt.Errorf("failed to initialize FCM client: %w", err)
	}

	return &NotificationService{
		notificationRepo: notificationRepo,
		userRepo:         userRepo,
		fcmClient:        fcmClient,
		emailService:     emailService,
		smsService:       smsService,
	}, nil
}

// RegisterFCMToken registers a user's FCM token
func (s *NotificationService) RegisterFCMToken(ctx context.Context, userID *primitive.ObjectID, token, platform string) error {
	fcmToken := &models.FCMToken{
		UserID:    userID,
		Token:     token,
		Platform:  platform,
		IsActive:  true,
		CreatedAt: time.Now(),
		UpdatedAt: time.Now(),
	}

	return s.notificationRepo.CreateFCMToken(ctx, fcmToken)
}

// SendOrderPlacedNotification sends order placed notification
func (s *NotificationService) SendOrderPlacedNotification(ctx context.Context, userID primitive.ObjectID, orderID string) error {
	notification := &models.Notification{
		UserID:    &userID,
		Title:     "Order Confirmed! 🎉",
		Body:      fmt.Sprintf("Your order #%s has been placed successfully. We'll notify you when it ships.", orderID),
		Type:      models.NotificationOrderPlaced,
		Data: map[string]interface{}{
			"type":     "order",
			"targetId": orderID,
		},
		IsRead:    false,
		IsSent:    false,
		CreatedAt: time.Now(),
		UpdatedAt: time.Now(),
	}

	return s.sendNotification(ctx, notification)
}

// SendOrderShippedNotification sends order shipped notification
func (s *NotificationService) SendOrderShippedNotification(ctx context.Context, userID primitive.ObjectID, orderID, trackingNumber string) error {
	notification := &models.Notification{
		UserID:    &userID,
		Title:     "Order Shipped! 📦",
		Body:      fmt.Sprintf("Your order #%s is on its way. Track it with: %s", orderID, trackingNumber),
		Type:      models.NotificationOrderShipped,
		Data: map[string]interface{}{
			"type":           "order",
			"targetId":       orderID,
			"trackingNumber": trackingNumber,
		},
		IsRead:    false,
		IsSent:    false,
		CreatedAt: time.Now(),
		UpdatedAt: time.Now(),
	}

	return s.sendNotification(ctx, notification)
}

// SendOrderDeliveredNotification sends order delivered notification
func (s *NotificationService) SendOrderDeliveredNotification(ctx context.Context, userID primitive.ObjectID, orderID string) error {
	notification := &models.Notification{
		UserID:    &userID,
		Title:     "Order Delivered! ✅",
		Body:      fmt.Sprintf("Your order #%s has been delivered. Enjoy your purchase!", orderID),
		Type:      models.NotificationOrderDelivered,
		Data: map[string]interface{}{
			"type":     "order",
			"targetId": orderID,
		},
		IsRead:    false,
		IsSent:    false,
		CreatedAt: time.Now(),
		UpdatedAt: time.Now(),
	}

	return s.sendNotification(ctx, notification)
}

// SendOrderCancelledNotification sends order cancelled notification
func (s *NotificationService) SendOrderCancelledNotification(ctx context.Context, userID primitive.ObjectID, orderID string) error {
	notification := &models.Notification{
		UserID:    &userID,
		Title:     "Order Cancelled",
		Body:      fmt.Sprintf("Your order #%s has been cancelled. Refund will be processed within 3-5 days.", orderID),
		Type:      models.NotificationOrderCancelled,
		Data: map[string]interface{}{
			"type":     "order",
			"targetId": orderID,
		},
		IsRead:    false,
		IsSent:    false,
		CreatedAt: time.Now(),
		UpdatedAt: time.Now(),
	}

	return s.sendNotification(ctx, notification)
}

// SendFlashSaleNotification sends flash sale notification
func (s *NotificationService) SendFlashSaleNotification(ctx context.Context, saleTitle string, discountPercent int) error {
	notification := &models.Notification{
		Title: "⚡ Flash Sale Alert!",
		Body:  fmt.Sprintf("%s - Get %d%% OFF! Limited time only.", saleTitle, discountPercent),
		Type:  models.NotificationFlashSale,
		Data: map[string]interface{}{
			"type":     "promotion",
			"targetId": "flash_sale",
		},
		IsRead:    false,
		IsSent:    false,
		CreatedAt: time.Now(),
		UpdatedAt: time.Now(),
	}

	return s.sendBroadcastNotification(ctx, notification)
}

// SendAbandonedCartNotification sends abandoned cart notification
func (s *NotificationService) SendAbandonedCartNotification(ctx context.Context, userID primitive.ObjectID, itemCount int, totalAmount float64) error {
	notification := &models.Notification{
		UserID:    &userID,
		Title:     "Items Still in Your Cart 🛒",
		Body:      fmt.Sprintf("You have %d items worth $%.2f waiting. Complete your purchase now!", itemCount, totalAmount),
		Type:      models.NotificationAbandonedCart,
		Data: map[string]interface{}{
			"type":     "cart",
			"targetId": "cart",
		},
		IsRead:    false,
		IsSent:    false,
		CreatedAt: time.Now(),
		UpdatedAt: time.Now(),
	}

	return s.sendNotification(ctx, notification)
}

// SendBackInStockNotification sends back in stock notification
func (s *NotificationService) SendBackInStockNotification(ctx context.Context, userID primitive.ObjectID, productName, productID string) error {
	notification := &models.Notification{
		UserID:    &userID,
		Title:     "Back in Stock! 🎯",
		Body:      fmt.Sprintf("%s is now available. Get it before it sells out again!", productName),
		Type:      models.NotificationBackInStock,
		Data: map[string]interface{}{
			"type":     "product",
			"targetId": productID,
		},
		IsRead:    false,
		IsSent:    false,
		CreatedAt: time.Now(),
		UpdatedAt: time.Now(),
	}

	return s.sendNotification(ctx, notification)
}

// SendLoyaltyPointsNotification sends loyalty points notification
func (s *NotificationService) SendLoyaltyPointsNotification(ctx context.Context, userID primitive.ObjectID, points int, tier string) error {
	notification := &models.Notification{
		UserID:    &userID,
		Title:     "You Earned Points! 🌟",
		Body:      fmt.Sprintf("+%d loyalty points added. You're now a %s member!", points, tier),
		Type:      models.NotificationLoyaltyPoints,
		Data: map[string]interface{}{
			"type":     "loyalty",
			"targetId": "rewards",
			"points":   points,
			"tier":     tier,
		},
		IsRead:    false,
		IsSent:    false,
		CreatedAt: time.Now(),
		UpdatedAt: time.Now(),
	}

	return s.sendNotification(ctx, notification)
}

// SendStreakBonusNotification sends streak bonus notification
func (s *NotificationService) SendStreakBonusNotification(ctx context.Context, userID primitive.ObjectID, streak, bonusPoints int) error {
	notification := &models.Notification{
		UserID:    &userID,
		Title:     "Login Streak Bonus! 🔥",
		Body:      fmt.Sprintf("%d day streak! You earned %d bonus points.", streak, bonusPoints),
		Type:      models.NotificationStreakBonus,
		Data: map[string]interface{}{
			"type":        "loyalty",
			"targetId":    "rewards",
			"streak":      streak,
			"bonusPoints": bonusPoints,
		},
		IsRead:    false,
		IsSent:    false,
		CreatedAt: time.Now(),
		UpdatedAt: time.Now(),
	}

	return s.sendNotification(ctx, notification)
}

// SendTierUpgradeNotification sends tier upgrade notification
func (s *NotificationService) SendTierUpgradeNotification(ctx context.Context, userID primitive.ObjectID, oldTier, newTier models.LoyaltyTier) error {
	newTierInfo := newTier.GetTierInfo()
	notification := &models.Notification{
		UserID:    &userID,
		Title:     fmt.Sprintf("Tier Upgrade! %s", newTierInfo.Icon),
		Body:      fmt.Sprintf("Congratulations! You've been upgraded to %s tier with exclusive benefits!", newTierInfo.Name),
		Type:      models.NotificationTierUpgrade,
		Data: map[string]interface{}{
			"type":     "loyalty",
			"targetId": "rewards",
			"oldTier":  string(oldTier),
			"newTier":  string(newTier),
		},
		IsRead:    false,
		IsSent:    false,
		CreatedAt: time.Now(),
		UpdatedAt: time.Now(),
	}

	return s.sendNotification(ctx, notification)
}

// sendNotification sends notification to a specific user
func (s *NotificationService) sendNotification(ctx context.Context, notification *models.Notification) error {
	// Save notification to database
	if err := s.notificationRepo.Create(ctx, notification); err != nil {
		return fmt.Errorf("failed to save notification: %w", err)
	}

	// Get user preferences if user is specified
	if notification.UserID != nil {
		prefs, err := s.notificationRepo.GetUserPreferences(ctx, *notification.UserID)
		if err != nil {
			log.Printf("Failed to get user preferences: %v", err)
			// Continue with default preferences
		}

		// Check if user has disabled this type of notification
		if prefs != nil && !s.isNotificationTypeEnabled(prefs, notification.Type) {
			return nil
		}

		// Get user's FCM tokens
		tokens, err := s.notificationRepo.GetActiveFCMTokens(ctx, notification.UserID)
		if err != nil {
			log.Printf("Failed to get FCM tokens for user %s: %v", notification.UserID.Hex(), err)
		} else {
			// Send push notification
			s.sendPushNotification(ctx, tokens, notification)
		}

		// Send email notification if enabled
		if prefs == nil || prefs.EnableEmailNotifications {
			s.sendEmailNotification(ctx, *notification.UserID, notification)
		}

		// Send SMS notification if enabled
		if prefs != nil && prefs.EnableSMSNotifications {
			s.sendSMSNotification(ctx, *notification.UserID, notification)
		}
	}

	return nil
}

// sendBroadcastNotification sends notification to all users
func (s *NotificationService) sendBroadcastNotification(ctx context.Context, notification *models.Notification) error {
	// Save notification to database
	if err := s.notificationRepo.Create(ctx, notification); err != nil {
		return fmt.Errorf("failed to save notification: %w", err)
	}

	// Get all active FCM tokens
	tokens, err := s.notificationRepo.GetAllActiveFCMTokens(ctx)
	if err != nil {
		log.Printf("Failed to get all FCM tokens: %v", err)
		return err
	}

	// Send push notification to all tokens
	return s.sendPushNotification(ctx, tokens, notification)
}

// sendPushNotification sends FCM push notification
func (s *NotificationService) sendPushNotification(ctx context.Context, tokens []models.FCMToken, notification *models.Notification) error {
	if len(tokens) == 0 {
		return nil
	}

	// Prepare FCM message
	message := &messaging.MulticastMessage{
		Notification: &messaging.Notification{
			Title: notification.Title,
			Body:  notification.Body,
		},
		Data: s.convertDataToStringMap(notification.Data),
		Android: &messaging.AndroidConfig{
			Notification: &messaging.AndroidNotification{
				ChannelID: s.getChannelID(notification.Type),
				Priority:  messaging.PriorityHigh,
			},
		},
		APNS: &messaging.APNSConfig{
			Payload: &messaging.APNSPayload{
				Aps: &messaging.Aps{
					Alert: &messaging.ApsAlert{
						Title: notification.Title,
						Body:  notification.Body,
					},
					Sound: "default",
				},
			},
		},
		Tokens: s.extractTokenStrings(tokens),
	}

	// Send multicast message
	response, err := s.fcmClient.SendMulticast(ctx, message)
	if err != nil {
		log.Printf("Failed to send FCM message: %v", err)
		return err
	}

	log.Printf("Successfully sent %d messages, %d failed", response.SuccessCount, response.FailureCount)

	// Handle failed tokens (remove invalid ones)
	if response.FailureCount > 0 {
		s.handleFailedTokens(ctx, tokens, response)
	}

	// Update notification as sent
	notification.IsSent = true
	now := time.Now()
	notification.SentAt = &now
	notification.UpdatedAt = now

	return s.notificationRepo.Update(ctx, notification)
}

// sendEmailNotification sends email notification
func (s *NotificationService) sendEmailNotification(ctx context.Context, userID primitive.ObjectID, notification *models.Notification) {
	if s.emailService == nil {
		return
	}

	user, err := s.userRepo.GetByID(ctx, userID)
	if err != nil {
		log.Printf("Failed to get user for email notification: %v", err)
		return
	}

	// Send email asynchronously
	go func() {
		if err := s.emailService.SendNotificationEmail(user.Email, notification.Title, notification.Body); err != nil {
			log.Printf("Failed to send email notification: %v", err)
		}
	}()
}

// sendSMSNotification sends SMS notification
func (s *NotificationService) sendSMSNotification(ctx context.Context, userID primitive.ObjectID, notification *models.Notification) {
	if s.smsService == nil {
		return
	}

	user, err := s.userRepo.GetByID(ctx, userID)
	if err != nil {
		log.Printf("Failed to get user for SMS notification: %v", err)
		return
	}

	// Send SMS asynchronously
	go func() {
		if err := s.smsService.SendNotificationSMS(user.Phone, notification.Body); err != nil {
			log.Printf("Failed to send SMS notification: %v", err)
		}
	}()
}

// Helper functions

func (s *NotificationService) isNotificationTypeEnabled(prefs *models.NotificationPreference, notifType models.NotificationType) bool {
	switch notifType {
	case models.NotificationOrderPlaced, models.NotificationOrderConfirmed,
		 models.NotificationOrderShipped, models.NotificationOrderDelivered,
		 models.NotificationOrderCancelled, models.NotificationOrderReturned,
		 models.NotificationOrderRefunded:
		return prefs.EnableOrderUpdates

	case models.NotificationFlashSale, models.NotificationCouponReminder,
		 models.NotificationSeasonalOffer, models.NotificationNewProduct,
		 models.NotificationPriceDiscount:
		return prefs.EnablePromotions

	case models.NotificationAbandonedCart, models.NotificationBackInStock,
		 models.NotificationWishlistDiscount, models.NotificationReviewReminder:
		return prefs.EnableReminders

	case models.NotificationLoyaltyPoints, models.NotificationStreakBonus,
		 models.NotificationTierUpgrade, models.NotificationVoucherExpiring:
		return prefs.EnableLoyaltyUpdates

	default:
		return true
	}
}

func (s *NotificationService) getChannelID(notifType models.NotificationType) string {
	switch notifType {
	case models.NotificationOrderPlaced, models.NotificationOrderConfirmed,
		 models.NotificationOrderShipped, models.NotificationOrderDelivered,
		 models.NotificationOrderCancelled, models.NotificationOrderReturned,
		 models.NotificationOrderRefunded:
		return "order_notifications"

	case models.NotificationFlashSale, models.NotificationCouponReminder,
		 models.NotificationSeasonalOffer, models.NotificationNewProduct,
		 models.NotificationPriceDiscount:
		return "promotional_notifications"

	case models.NotificationAbandonedCart, models.NotificationBackInStock,
		 models.NotificationWishlistDiscount, models.NotificationReviewReminder,
		 models.NotificationLoyaltyPoints, models.NotificationStreakBonus,
		 models.NotificationTierUpgrade, models.NotificationVoucherExpiring:
		return "behavioral_notifications"

	default:
		return "order_notifications"
	}
}

func (s *NotificationService) convertDataToStringMap(data map[string]interface{}) map[string]string {
	stringMap := make(map[string]string)
	for key, value := range data {
		switch v := value.(type) {
		case string:
			stringMap[key] = v
		case int:
			stringMap[key] = strconv.Itoa(v)
		case float64:
			stringMap[key] = strconv.FormatFloat(v, 'f', -1, 64)
		case bool:
			stringMap[key] = strconv.FormatBool(v)
		default:
			// Convert to JSON string for complex types
			if jsonBytes, err := json.Marshal(v); err == nil {
				stringMap[key] = string(jsonBytes)
			}
		}
	}
	return stringMap
}

func (s *NotificationService) extractTokenStrings(tokens []models.FCMToken) []string {
	var tokenStrings []string
	for _, token := range tokens {
		tokenStrings = append(tokenStrings, token.Token)
	}
	return tokenStrings
}

func (s *NotificationService) handleFailedTokens(ctx context.Context, tokens []models.FCMToken, response *messaging.BatchResponse) {
	for i, resp := range response.Responses {
		if !resp.Success && i < len(tokens) {
			// Mark token as inactive if it's invalid
			if resp.Error != nil && (resp.Error.ErrorCode == "registration-token-not-registered" ||
				resp.Error.ErrorCode == "invalid-registration-token") {
				if err := s.notificationRepo.DeactivateFCMToken(ctx, tokens[i].Token); err != nil {
					log.Printf("Failed to deactivate invalid token: %v", err)
				}
			}
		}
	}
}

// API methods for notification management

// GetUserNotifications gets notifications for a user
func (s *NotificationService) GetUserNotifications(ctx context.Context, userID primitive.ObjectID, limit, offset int, unreadOnly bool) ([]models.Notification, error) {
	return s.notificationRepo.GetUserNotifications(ctx, userID, limit, offset, unreadOnly)
}

// MarkNotificationAsRead marks a notification as read
func (s *NotificationService) MarkNotificationAsRead(ctx context.Context, notificationID primitive.ObjectID, userID primitive.ObjectID) error {
	return s.notificationRepo.MarkAsRead(ctx, notificationID, userID)
}

// MarkAllNotificationsAsRead marks all notifications as read for a user
func (s *NotificationService) MarkAllNotificationsAsRead(ctx context.Context, userID primitive.ObjectID) error {
	return s.notificationRepo.MarkAllAsRead(ctx, userID)
}

// GetNotificationPreferences gets user's notification preferences
func (s *NotificationService) GetNotificationPreferences(ctx context.Context, userID primitive.ObjectID) (*models.NotificationPreference, error) {
	prefs, err := s.notificationRepo.GetUserPreferences(ctx, userID)
	if err != nil {
		// Create default preferences if not found
		prefs = models.DefaultNotificationPreferences(userID)
		if createErr := s.notificationRepo.CreateUserPreferences(ctx, prefs); createErr != nil {
			return nil, fmt.Errorf("failed to create default preferences: %w", createErr)
		}
	}
	return prefs, nil
}

// UpdateNotificationPreferences updates user's notification preferences
func (s *NotificationService) UpdateNotificationPreferences(ctx context.Context, prefs *models.NotificationPreference) error {
	prefs.UpdatedAt = time.Now()
	return s.notificationRepo.UpdateUserPreferences(ctx, prefs)
}

// SendCampaign sends a notification campaign
func (s *NotificationService) SendCampaign(ctx context.Context, campaign *models.NotificationCampaign) error {
	// Update campaign status
	campaign.Status = models.CampaignStatusSending
	campaign.UpdatedAt = time.Now()

	if err := s.notificationRepo.UpdateCampaign(ctx, campaign); err != nil {
		return fmt.Errorf("failed to update campaign status: %w", err)
	}

	// Get target users based on audience criteria
	userIDs, err := s.getTargetUsers(ctx, campaign.TargetAudience)
	if err != nil {
		campaign.Status = models.CampaignStatusFailed
		s.notificationRepo.UpdateCampaign(ctx, campaign)
		return fmt.Errorf("failed to get target users: %w", err)
	}

	campaign.Recipients = len(userIDs)

	// Send notifications to all target users
	sentCount := 0
	for _, userID := range userIDs {
		notification := &models.Notification{
			UserID:    &userID,
			Title:     campaign.Title,
			Body:      campaign.Body,
			Type:      campaign.Type,
			Data:      campaign.Data,
			IsRead:    false,
			IsSent:    false,
			CreatedAt: time.Now(),
			UpdatedAt: time.Now(),
		}

		if err := s.sendNotification(ctx, notification); err != nil {
			log.Printf("Failed to send campaign notification to user %s: %v", userID.Hex(), err)
		} else {
			sentCount++
		}
	}

	// Update campaign with results
	campaign.Status = models.CampaignStatusSent
	campaign.DeliveredCount = sentCount
	now := time.Now()
	campaign.SentAt = &now
	campaign.UpdatedAt = now

	return s.notificationRepo.UpdateCampaign(ctx, campaign)
}

// getTargetUsers gets user IDs based on campaign audience criteria
func (s *NotificationService) getTargetUsers(ctx context.Context, audience models.CampaignAudience) ([]primitive.ObjectID, error) {
	if audience.AllUsers {
		return s.userRepo.GetAllUserIDs(ctx)
	}

	// Implement specific audience targeting logic here
	// This would involve querying users based on various criteria
	// For now, return specific user IDs if provided
	if len(audience.SpecificUserIDs) > 0 {
		return audience.SpecificUserIDs, nil
	}

	// Default to empty list
	return []primitive.ObjectID{}, nil
}

File: lib/services/notification_service.dart
import 'dart:convert';
import 'dart:io';
import 'package:flutter_local_notifications/flutter_local_notifications.dart';
import 'package:flutter/material.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter_secure_storage/flutter_secure_storage.dart';
import 'api_service.dart';

class NotificationService {
  static final NotificationService _instance = NotificationService._();
  factory NotificationService() => _instance;
  NotificationService._();

  final FlutterLocalNotificationsPlugin _localNotifications = FlutterLocalNotificationsPlugin();
  static const FlutterSecureStorage _storage = FlutterSecureStorage();

  // Notification channels
  static const String orderChannel = 'order_notifications';
  static const String promotionalChannel = 'promotional_notifications';
  static const String behavioralChannel = 'behavioral_notifications';

  Future<void> initialize() async {
    try {
      // Initialize local notifications
      await _initializeLocalNotifications();

      debugPrint('Notification service initialized successfully');
    } catch (e) {
      debugPrint('Notification service initialization failed: $e');
    }
  }

  Future<void> _requestPermission() async {
    if (!kIsWeb && Platform.isAndroid) {
      await _localNotifications
          .resolvePlatformSpecificImplementation<AndroidFlutterLocalNotificationsPlugin>()
          ?.requestNotificationsPermission();
    }
  }

  Future<void> _initializeLocalNotifications() async {
    const androidSettings = AndroidInitializationSettings('@mipmap/ic_launcher');
    const iosSettings = DarwinInitializationSettings(
      requestAlertPermission: true,
      requestBadgePermission: true,
      requestSoundPermission: true,
    );

    const initSettings = InitializationSettings(
      android: androidSettings,
      iOS: iosSettings,
    );

    await _localNotifications.initialize(
      initSettings,
      onDidReceiveNotificationResponse: _handleNotificationTap,
    );

    // Create notification channels for Android
    if (!kIsWeb && Platform.isAndroid) {
      await _createNotificationChannels();
    }

    // Request permissions
    await _requestPermission();
  }

  Future<void> _createNotificationChannels() async {
    const orderChannelAndroid = AndroidNotificationChannel(
      orderChannel,
      'Order Updates',
      description: 'Notifications about your orders',
      importance: Importance.high,
    );

    const promotionalChannelAndroid = AndroidNotificationChannel(
      promotionalChannel,
      'Promotions & Offers',
      description: 'Special deals and promotional offers',
      importance: Importance.defaultImportance,
    );

    const behavioralChannelAndroid = AndroidNotificationChannel(
      behavioralChannel,
      'Reminders & Alerts',
      description: 'Cart reminders and stock alerts',
      importance: Importance.defaultImportance,
    );

    await _localNotifications
        .resolvePlatformSpecificImplementation<AndroidFlutterLocalNotificationsPlugin>()
        ?.createNotificationChannel(orderChannelAndroid);

    await _localNotifications
        .resolvePlatformSpecificImplementation<AndroidFlutterLocalNotificationsPlugin>()
        ?.createNotificationChannel(promotionalChannelAndroid);

    await _localNotifications
        .resolvePlatformSpecificImplementation<AndroidFlutterLocalNotificationsPlugin>()
        ?.createNotificationChannel(behavioralChannelAndroid);
  }

  // Handle notification tap
  void _handleNotificationTap(NotificationResponse response) {
    if (response.payload != null) {
      final data = json.decode(response.payload!);
      _navigateToScreen(data);
    }
  }

  String _getChannelId(String? type) {
    switch (type) {
      case 'order':
        return orderChannel;
      case 'promotional':
        return promotionalChannel;
      case 'behavioral':
        return behavioralChannel;
      default:
        return orderChannel;
    }
  }

  void _navigateToScreen(Map<String, dynamic> data) {
    final type = data['type'];
    final targetId = data['targetId'];

    switch (type) {
      case 'order':
        // Navigate to order details
        // Navigator.pushNamed(context, '/order/$targetId');
        break;
      case 'product':
        // Navigate to product details
        // Navigator.pushNamed(context, '/product/$targetId');
        break;
      case 'promotion':
        // Navigate to promotions
        // Navigator.pushNamed(context, '/promotions');
        break;
      case 'cart':
        // Navigate to cart
        // Navigator.pushNamed(context, '/cart');
        break;
    }
  }

  // Send different types of notifications

  // Transactional Notifications
  Future<void> sendOrderPlacedNotification(String orderId) async {
    await _sendLocalNotification(
      title: 'Order Confirmed! 🎉',
      body: 'Your order #$orderId has been placed successfully. We\'ll notify you when it ships.',
      data: {'type': 'order', 'targetId': orderId},
      channel: orderChannel,
    );
  }

  Future<void> sendOrderShippedNotification(String orderId, String trackingNumber) async {
    await _sendLocalNotification(
      title: 'Order Shipped! 📦',
      body: 'Your order #$orderId is on its way. Track it with: $trackingNumber',
      data: {'type': 'order', 'targetId': orderId},
      channel: orderChannel,
    );
  }

  Future<void> sendOrderDeliveredNotification(String orderId) async {
    await _sendLocalNotification(
      title: 'Order Delivered! ✅',
      body: 'Your order #$orderId has been delivered. Enjoy your purchase!',
      data: {'type': 'order', 'targetId': orderId},
      channel: orderChannel,
    );
  }

  Future<void> sendOrderCancelledNotification(String orderId) async {
    await _sendLocalNotification(
      title: 'Order Cancelled',
      body: 'Your order #$orderId has been cancelled. Refund will be processed within 3-5 days.',
      data: {'type': 'order', 'targetId': orderId},
      channel: orderChannel,
    );
  }

  // Promotional Notifications
  Future<void> sendFlashSaleNotification(String saleTitle, int discountPercent) async {
    await _sendLocalNotification(
      title: '⚡ Flash Sale Alert!',
      body: '$saleTitle - Get $discountPercent% OFF! Limited time only.',
      data: {'type': 'promotion', 'targetId': 'flash_sale'},
      channel: promotionalChannel,
    );
  }

  Future<void> sendCouponReminderNotification(String couponCode, int daysLeft) async {
    await _sendLocalNotification(
      title: 'Coupon Expiring Soon! ⏰',
      body: 'Your coupon $couponCode expires in $daysLeft days. Use it before it\'s gone!',
      data: {'type': 'promotion', 'targetId': couponCode},
      channel: promotionalChannel,
    );
  }

  Future<void> sendSeasonalOfferNotification(String season, int discount) async {
    await _sendLocalNotification(
      title: '$season Sale Now Live! 🎊',
      body: 'Celebrate with up to $discount% OFF on selected items.',
      data: {'type': 'promotion', 'targetId': 'seasonal'},
      channel: promotionalChannel,
    );
  }

  // Behavioral Notifications
  Future<void> sendAbandonedCartNotification(int itemCount, double totalAmount) async {
    await _sendLocalNotification(
      title: 'Items Still in Your Cart 🛒',
      body: 'You have $itemCount items worth \$$totalAmount waiting. Complete your purchase now!',
      data: {'type': 'cart', 'targetId': 'cart'},
      channel: behavioralChannel,
    );
  }

  Future<void> sendBackInStockNotification(String productName, String productId) async {
    await _sendLocalNotification(
      title: 'Back in Stock! 🎯',
      body: '$productName is now available. Get it before it sells out again!',
      data: {'type': 'product', 'targetId': productId},
      channel: behavioralChannel,
    );
  }

  Future<void> sendLoyaltyPointsNotification(int points, String tier) async {
    await _sendLocalNotification(
      title: 'You Earned Points! 🌟',
      body: '+$points loyalty points added. You\'re now a $tier member!',
      data: {'type': 'loyalty', 'targetId': 'rewards'},
      channel: behavioralChannel,
    );
  }

  Future<void> sendStreakBonusNotification(int streak, int bonusPoints) async {
    await _sendLocalNotification(
      title: 'Login Streak Bonus! 🔥',
      body: '$streak day streak! You earned $bonusPoints bonus points.',
      data: {'type': 'loyalty', 'targetId': 'rewards'},
      channel: behavioralChannel,
    );
  }

  // Helper method to send local notifications
  Future<void> _sendLocalNotification({
    required String title,
    required String body,
    required Map<String, dynamic> data,
    required String channel,
  }) async {
    final androidDetails = AndroidNotificationDetails(
      channel,
      channel,
      importance: Importance.high,
      priority: Priority.high,
      showWhen: true,
    );

    const iosDetails = DarwinNotificationDetails(
      presentAlert: true,
      presentBadge: true,
      presentSound: true,
    );

    final details = NotificationDetails(
      android: androidDetails,
      iOS: iosDetails,
    );

    await _localNotifications.show(
      DateTime.now().millisecondsSinceEpoch ~/ 1000,
      title,
      body,
      details,
      payload: json.encode(data),
    );
  }

  // Update notification preferences
  Future<void> updateNotificationPreferences({
    bool orderUpdates = true,
    bool promotions = true,
    bool reminders = true,
    bool loyaltyRewards = true,
    bool backInStock = true,
    bool flashSales = true,
  }) async {
    try {
      // Save preferences securely
      await _storage.write(key: 'notifications_order_updates', value: orderUpdates.toString());
      await _storage.write(key: 'notifications_promotions', value: promotions.toString());
      await _storage.write(key: 'notifications_reminders', value: reminders.toString());
      await _storage.write(key: 'notifications_loyalty_rewards', value: loyaltyRewards.toString());
      await _storage.write(key: 'notifications_back_in_stock', value: backInStock.toString());
      await _storage.write(key: 'notifications_flash_sales', value: flashSales.toString());

      // Sync with backend
      await _syncPreferencesWithBackend({
        'orderUpdates': orderUpdates,
        'promotions': promotions,
        'reminders': reminders,
        'loyaltyRewards': loyaltyRewards,
        'backInStock': backInStock,
        'flashSales': flashSales,
      });
    } catch (e) {
      debugPrint('Error updating notification preferences: $e');
    }
  }

  // Get notification preferences
  Future<Map<String, bool>> getNotificationPreferences() async {
    try {
      return {
        'orderUpdates': (await _storage.read(key: 'notifications_order_updates')) == 'true',
        'promotions': (await _storage.read(key: 'notifications_promotions')) == 'true',
        'reminders': (await _storage.read(key: 'notifications_reminders')) == 'true',
        'loyaltyRewards': (await _storage.read(key: 'notifications_loyalty_rewards')) == 'true',
        'backInStock': (await _storage.read(key: 'notifications_back_in_stock')) == 'true',
        'flashSales': (await _storage.read(key: 'notifications_flash_sales')) == 'true',
      };
    } catch (e) {
      debugPrint('Error getting notification preferences: $e');
      return {
        'orderUpdates': true,
        'promotions': true,
        'reminders': true,
        'loyaltyRewards': true,
        'backInStock': true,
        'flashSales': true,
      };
    }
  }

  // Sync notification preferences with backend
  Future<void> _syncPreferencesWithBackend(Map<String, bool> preferences) async {
    try {
      await ApiService.updateNotificationPreferences(preferences: preferences);
      debugPrint('Notification preferences synced with backend successfully');
    } catch (e) {
      debugPrint('Error syncing notification preferences with backend: $e');
    }
  }

  // Get user notifications from backend
  Future<List<Map<String, dynamic>>> getUserNotifications({
    int page = 1,
    int limit = 20,
    bool unreadOnly = false,
  }) async {
    try {
      final response = await ApiService.getUserNotifications(
        page: page,
        limit: limit,
        unreadOnly: unreadOnly,
      );
      return List<Map<String, dynamic>>.from(response['data']['notifications'] ?? []);
    } catch (e) {
      debugPrint('Error fetching user notifications: $e');
      return [];
    }
  }

  // Mark notification as read
  Future<void> markNotificationAsRead(String notificationId) async {
    try {
      await ApiService.markNotificationAsRead(notificationId: notificationId);
    } catch (e) {
      debugPrint('Error marking notification as read: $e');
    }
  }

  // Mark all notifications as read
  Future<void> markAllNotificationsAsRead() async {
    try {
      await ApiService.markAllNotificationsAsRead();
    } catch (e) {
      debugPrint('Error marking all notifications as read: $e');
    }
  }
}